src/stackwalk.c

```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
     jlbacktrace();
 }
 
+JL_DLLEXPORT void jl_print_task_backtraces(void)
+{
+    jl_task_t *t;
+    jl_ptls_t ptls;
+    arraylist_t *live_tasks;
+    size_t i, j, n;
+
+    for (i = 0; i < jl_n_threads; i++) {
+        ptls = jl_all_tls_states[i];
```

Comment:
traditionally, we call this `ptls2` in gC, but you also aren't allowed to access that outside of GC, since it is thread-local state for a different thread.

Reply:
Yes, it will be racey, but will it simply fail on some platforms?

Reply:
If, on some platforms, it is not possible to access another thread's space through `jl_all_tls_states`, the only way around I can see is to do a GC stop-the-world and have each thread print its own `live_tasks` which will no longer be racy, but make this a lot more complicated and slow.

Reply:
A thread, in general, cannot print the stack of another thread's task either, so having it run behind the `jl_gc_wait_for_the_world` logic (running on every thread) would be useful. Though deadlock reporting doesn't precisely need that, since that may be only essential to run if we have run out of all other useful work and halted the other threads anyways.

Reply:
The `live_tasks` pointer may move from under you while you are looking at it, making this very dangerous. You are right that sometimes we examine values in the `jl_all_tls_states`, but only if they are marked atomic or are protected by a lock

Reply:
As far as I know, we only intended to call this from `gdb`, in which case the world will already be artificially stopped.

I agree with you both, that I don't think this is safe to call from inside the julia runtime.

I think we should make sure that the function comments make it clear that this isn't safe to call while the runtime is running.
It might make sense to provide another version of it, which force-pauses all the threads before printing, reusing the GC machinery, but that seems like a follow up PR?

Reply:
FWIW the Go version of this stops the world so it can be safely called from 'userspace': https://github.com/golang/go/blob/6d8ec893039a39f495c8139012e47754e4518b70/src/runtime/mprof.go#L1193-L1196

This is sometimes put in an HTTP handler for easy access; would be nice for us to eventually be able to do the same. GDB is a good start ofc :)